garage/tf/algos/dqn.py

```diff
@@ -0,0 +1,282 @@
+                    next_obs = np.asarray(next_obs)
+                self.replay_buffer.add_transitions(
+                    observation=[obs],
+                    action=[action],
+                    reward=[reward],
+                    terminal=[done],
+                    next_observation=[next_obs],
+                )
+
+                episode_rewards[-1] += reward
+                ts += 1
+                obs = next_obs
+
+                if done:
+                    episode_length.append(ts)
+                    ts = 0
+                    episode_rewards.append(0.)
+                    obs = self.env.reset()
+                    if self.use_atari_wrappers:
+                        obs = np.asarray(obs)
+
+                if self.replay_buffer.n_transitions_stored >= self.min_buffer_size:  # noqa: E501
+                    self.evaluate = True
+                    qf_loss = self.optimize_policy(itr, None)
+                    episode_qf_losses.append(qf_loss)
+                    if itr % self.target_network_update_freq == 0:
+                        self.sess.run(self._qf_update_ops, feed_dict=dict())
+
+                if self.plot:
+                    self.plotter.update_plot(self.policy, self.max_path_length)
+                    if self.pause_for_plot:
+                        input("Plotting evaluation run: Press Enter to "
+                              "continue...")
+
+                if self.evaluate:
+                    mean100ep_rewards = round(
+                        np.mean(episode_rewards[-100:]), 1)
+                    mean100ep_qf_loss = np.mean(episode_qf_losses[-100:])
+                    if itr % self.print_freq == 0:
+                        logger.record_tabular('Iteration', itr)
+                        logger.record_tabular('AverageReturn',
+                                              mean100ep_rewards)
+                        logger.record_tabular('StdReturn',
+                                              np.std(episode_rewards))
+                        logger.record_tabular('QFunction/AverageQFunctionLoss',
+                                              mean100ep_qf_loss)
+                        logger.record_tabular('EpisodeLength',
+                                              np.mean(episode_length))
+
+                logger.dump_tabular(with_prefix=False)
+
+        if created_session:
+            self.sess.close()
+
+        return mean100ep_rewards
+
+    @overrides
+    def optimize_policy(self, itr, sample_data):
+        """Optimize network using experiences from replay buffer."""
+        transitions = self.replay_buffer.sample(self.buffer_batch_size)
+
+        observations = transitions["observation"]
+        rewards = transitions["reward"]
+        actions = transitions["action"]
+        next_observations = transitions["next_observation"]
+        dones = transitions["terminal"]
+
+        if len(self.env.spec.observation_space.shape) == 3:
```

Comment:
emm, I feel it is strange to have an if clause here. Can we either include it in a util function?